slowmovie.py

```diff
@@ -10,205 +10,233 @@
+        currentVideo = os.path.join(viddir, random.choice(videos))
+
+# ...then the last played file...
+if not currentVideo and os.path.isfile("nowPlaying"):
+    with open("nowPlaying") as file:
+        lastVideo = file.readline().strip()
+        if os.path.isfile(lastVideo):
+            currentVideo = lastVideo
         else:
-            currentVideo = movieList[0]
+            os.remove("nowPlaying")
+
+# ...then we look in the videos folder.
+if not currentVideo:
+    videos = os.listdir(viddir)
+    for file in videos:
+        if supported_filetype(file):
+            currentVideo = os.path.join(viddir, file)
+            break
+
+# If none of the above worked, exit.
+if not currentVideo:
+    print("No videos found")
+    sys.exit()
+
+print("Update interval: " + str(args.delay))
+if not args.random:
+    print("Frame increment: " + str(args.increment))
+
+with open("nowPlaying", "w") as file:
+    file.write(os.path.abspath(currentVideo))
+
+videoFilename = os.path.basename(currentVideo)
+
+if not args.loop:
+    viddir = os.path.dirname(currentVideo)
+    videos = list(filter(supported_filetype, os.listdir(viddir)))
+    fileIndex = videos.index(videoFilename)
+
+logfile = os.path.join(logdir, videoFilename + ".progress")
+
+epd = epd_driver.EPD()
+width = epd.width
+height = epd.height
+
+frameCount, framerate, frametime = video_info(currentVideo)
+
+if not args.random:
+    if args.start:
+        currentFrame = clamp(args.start, 0, frameCount)
+        print("Starting at frame " + str(currentFrame))
+    elif (os.path.isfile(logfile)):
+        # Read current frame from logfile
+        with open(logfile) as log:
+            try:
+                currentFrame = int(log.readline())
+                currentFrame = clamp(currentFrame, 0, frameCount)
+            except ValueError:
+                currentFrame = 0
+    else:
+        currentFrame = 0
+
+lastVideo = None
+
+while 1:
+    if lastVideo != currentVideo:
+        print(f"Playing '{videoFilename}'")
+        lastVideo = currentVideo
+
+    timeStart = time.perf_counter()
+    epd.init()
 
-    log = open(logdir + '%s<progress'%currentVideo, 'w')
-    log.write(str(currentPosition))
-    log.close() 
+    if args.random:
+        currentFrame = random.randint(0, frameCount)
 
+    msTimecode = "%dms" % (currentFrame * frametime)
 
-    f = open('nowPlaying', 'w')
-    f.write(currentVideo)
-    f.close() 
-    
+    # Use ffmpeg to extract a frame from the movie, letterbox/pillarbox it and save it as frame.bmp
+    generate_frame(currentVideo, "/dev/shm/frame.bmp", msTimecode)
 
-#     epd.sleep()
-    time.sleep(frameDelay)
-    epd.init()
+    # Open frame.bmp in PIL
+    pil_im = Image.open("/dev/shm/frame.bmp")
 
+    if args.contrast != 1:
+        enhancer = ImageEnhance.Contrast(pil_im)
+        pil_im = enhancer.enhance(args.contrast)
 
+    # Dither the image into a 1 bit bitmap
+    #pil_im = pil_im.convert(mode = "1", dither = Image.FLOYDSTEINBERG)
```

Comment:
This exists in other versions of SlowMovie - is it not needed?

Reply:
The driver already does this.https://github.com/TomWhitwell/SlowMovie/blob/e66293e204ef933f284a5d8c8cdcee7fb99e0594/e-paper/RaspberryPi%26JetsonNano/python/lib/waveshare_epd/epd7in5_V2.py#L118
(Floyd-steinberg is the [default](https://pillow.readthedocs.io/en/stable/reference/Image.html#PIL.Image.Image.convert), so that parameter isn't necessary)